src/mlpack/core/data/image_bounding_box_impl.hpp

```diff
@@ -0,0 +1,134 @@
+/**
+ * @file core/data/image_bounding_box_impl.hpp
+ * @author Andrew Furey
+ *
+ * Draw bounding boxes and labels onto images.
+ *
+ * mlpack is free software; you may redistribute it and/or modify it under the
+ * terms of the 3-clause BSD license.  You should have received a copy of the
+ * 3-clause BSD license along with mlpack.  If not, see
+ * http://www.opensource.org/licenses/BSD-3-Clause for more information.
+ */
+
+#ifndef MLPACK_CORE_DATA_IMAGE_BOUNDING_BOX_IMPL_HPP
+#define MLPACK_CORE_DATA_IMAGE_BOUNDING_BOX_IMPL_HPP
+
+#include "image_bounding_box.hpp"
+#include "image_options.hpp"
+
+namespace mlpack {
+namespace data {
+
+template <typename MatType>
+inline void UpdatePixel(MatType& src,
+  const ImageInfo& srcOpt,
+  const size_t x,
+  const size_t y,
+  const typename MatType::elem_type red,
+  const typename MatType::elem_type green,
+  const typename MatType::elem_type blue)
+{
+  const size_t redChannel =
+    x * srcOpt.Channels() + y * srcOpt.Channels() * srcOpt.Width();
+
+  src.at(redChannel, 0) = red;
+  src.at(redChannel + 1, 0) = green;
+  src.at(redChannel + 2, 0) = blue;
+}
+
+template <typename MatType>
+inline void DrawLetter(MatType& src,
+  const ImageInfo& srcOpt,
+  const char letter,
+  const typename MatType::elem_type x,
+  const typename MatType::elem_type y,
+  const size_t size)
+{
+  const size_t fontWidth = 8;
+  for (size_t i = 0; i < fontWidth; i++)
+  {
+    for (size_t j = 0; j < fontWidth; j++)
+    {
+      const bool set = !(font8x8_basic[letter][i] & (unsigned char)(1 << j));
+      for (size_t k = 0; k < size * size; k++)
+      {
+        const size_t px = x + (j * size) + (k % size);
+        const size_t py = y + (i * size) + (k / size);
+        UpdatePixel(src, srcOpt, px, py, set, set, set);
+      }
+    }
+  }
+}
+
+template <typename ImageType, typename BoundingBoxesType>
+inline void BoundingBoxImage(ImageType& src,
+  const ImageInfo& srcOpt,
+  const BoundingBoxesType& bbox,
+  const std::string& className,
+  const typename ImageType::elem_type red,
+  const typename ImageType::elem_type green,
+  const typename ImageType::elem_type blue,
```

Comment:
For generality it might be better to accept a vector here, and check that its number of elements matches the number of channels of the image.